packages/block-library/src/embed/edit.js

```diff
@@ -0,0 +1,382 @@
+
+			this.props.setAttributes(
+				{
+					allowResponsive: ! allowResponsive,
+					className: classnames( className, responsiveClassNames ),
+				}
+			);
+		}
+
+		render() {
+			const { url, editingURL } = this.state;
+			const { caption, type, allowResponsive } = this.props.attributes;
+			const { fetching, setAttributes, isSelected, className, preview, cannotEmbed, supportsResponsive } = this.props;
+			const controls = (
+				<Fragment>
+					<BlockControls>
+						<Toolbar>
+							{ preview && ! cannotEmbed && (
+								<IconButton
+									className="components-toolbar__control"
+									label={ __( 'Edit URL' ) }
+									icon="edit"
+									onClick={ this.switchBackToURLInput }
+								/>
+							) }
+						</Toolbar>
+					</BlockControls>
+					{ supportsResponsive && (
+						<InspectorControls>
+							<PanelBody title={ __( 'Media Settings' ) } className="blocks-responsive">
+								<ToggleControl
+									label={ __( 'Automatically scale content' ) }
+									checked={ allowResponsive }
+									help={ this.getResponsiveHelp }
+									onChange={ this.toggleResponsive }
+								/>
+							</PanelBody>
+						</InspectorControls>
+					) }
+				</Fragment>
+			);
+
+			if ( fetching ) {
+				return (
+					<div className="wp-block-embed is-loading">
+						<Spinner />
+						<p>{ __( 'Embedding…' ) }</p>
+					</div>
+				);
+			}
+
+			// translators: %s: type of embed e.g: "YouTube", "Twitter", etc. "Embed" is used when no specific type exists
+			const label = sprintf( __( '%s URL' ), title );
+
+			// No preview, or we can't embed the current URL, or we've clicked the edit button.
+			if ( ! preview || cannotEmbed || editingURL ) {
+				return (
+					<Placeholder icon={ <BlockIcon icon={ icon } showColors /> } label={ label } className="wp-block-embed">
+						<form onSubmit={ this.setUrl }>
+							<input
+								type="url"
+								value={ url || '' }
+								className="components-placeholder__input"
+								aria-label={ label }
+								placeholder={ __( 'Enter URL to embed here…' ) }
+								onChange={ ( event ) => this.setState( { url: event.target.value } ) } />
+							<Button
+								isLarge
+								type="submit">
+								{ __( 'Embed' ) }
+							</Button>
+							{ cannotEmbed && <p className="components-placeholder__error">{ __( 'Sorry, we could not embed that content.' ) }</p> }
+						</form>
+					</Placeholder>
+				);
+			}
+
+			const html = 'photo' === type ? this.getPhotoHtml( preview ) : preview.html;
+			const { scripts } = preview;
+			const parsedUrl = parse( url );
+			const cannotPreview = includes( HOSTS_NO_PREVIEWS, parsedUrl.host.replace( /^www\./, '' ) );
+			// translators: %s: host providing embed content e.g: www.youtube.com
+			const iframeTitle = sprintf( __( 'Embedded content from %s' ), parsedUrl.host );
+			const sandboxClassnames = classnames( type, className );
+			const embedWrapper = 'wp-embed' === type ? (
```

Comment:
embedWrapper is only rendered when `cannotPreview` is false, and yet the elements are still created here.

Reply:
Probably makes sense to make the preview its own component so that it can handle that logic and error messages internally.